### DISCUSSION TITLE
Is there a way to keep the 'audioprocess' event running when page is not visible?

### DISCUSSION THREAD

**Discussion by RiccardoNL** [katspaugh/wavesurfer.js#2987] [Q&A]
Hi all,

I am using WaveSurfer for an internal radio project. I am using the 'audioprocess' event to tell if it is time to play the next item in the playlist (I'm using cues I programmed myself). However, when the page is not visible, the audioprocess event seems to not be transmitting any data. The page must be visible at all time, it seems. I am using MediaElement as a backend.

Is there a way to keep the 'audioprocess' event running when page is not visible? Am I missing something?

Kind regards,
Riccardo

**Comment by katspaugh** [comment]
You should try the timeupdate event instead, it’s emitted directly from the media element so it will work in the background too.
Be aware that its resolution is lower than audioprocess though, so your cueing might be less precise if the tab is not focused.
In v7, timeupdate is emitted from the wavesurfer object as well.

**Reply by RiccardoNL** [reply]
Just a second after I found this out, you commented. It seems to work. I’m having the player in a test rotation to see if this could also work 24/7 (I’m building a radio automation for myself). I’ll have this running for some days, and report back to you if this indeed was the solution :).

I give you a big compliment for how well WaveSurfer is built and how speedy your replies are. Amazing!

**Comment by RiccardoNL** [comment]
@katspaugh I see the timeupdate event updates only irregularly. This results in only being able to read the currentTime in seconds. Is there a way to make it update more regularly so I also can use milliseconds?

**Reply by katspaugh** [reply]
It's by design, see https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/timeupdate_event

> The event frequency is dependent on the system load, but will be thrown between about 4Hz and 66Hz (assuming the event handlers don't take longer than 250ms to run). User agents are encouraged to vary the frequency of the event based on the system load and the average cost of processing the event each time, so that the UI updates are not any more frequent than the user agent can comfortably handle while decoding the video.

**Reply by RiccardoNL** [reply]
Thanks. That's unfortunate. Currently I'm testing a different approach: a heartbeat server running in the background, sending messages over a web socket server. It is requesting the currentTime from the backend all the time, until it meets a cue point. Let's see how long this will work :).

(If this doesn't work for a long time, I'll accept the fact I can only use full seconds and not milliseconds).

**Reply by katspaugh** [reply]
You might be able to do something similar with Service Workers probably.

**Reply by RiccardoNL** [reply]
Thanks, @katspaugh. I will have a look.  I see this requires extra secured connections, so I first need to do some reading up on how to make trusted network connections, using vHosts and how to sign SSL certificates. I will let you know later if service workers were the solution.
